### PR TITLE
Add single retest functionality in Test/Edit

### PR DESCRIPTION
**Pull request by PetarNeshkov**
Closes: https://github.com/SoftUni-Internal/exam-systems-issues/issues/855
Closes: https://github.com/SoftUni-Internal/exam-systems-issues/issues/922

**Summary of the changes made**:
1. Implemented retesting for a single test in Test/Edit page
2. Fixed an issue which occurs with nullable checkboxes values

### CODE REVIEW

File: Services/Administration/OJS.Services.Administration.Business/Implementations/ProblemsBusinessService.cs [hunk 1]
```diff
@@ -54,16 +58,32 @@ public class ProblemsBusinessService : IProblemsBusinessService
...
+            if (retestBySingleTest)
+            {
+                problemIdFromQuery = await this.testData.GetProblemIdByTestId(id);
+
+                submissionsQueryable = this.submissionsData.GetAllByProblem(problemIdFromQuery);
+            }
+            else
+            {
+                submissionsQueryable = this.submissionsData.GetAllByProblem(id);
+            }
```
Comment (gogo4ds) [inline Services/Administration/OJS.Services.Administration.Business/Implementations/ProblemsBusinessService.cs:79]: This is veery misleading. You are essentially using the parameter `id` to mean two different things - in one case it is the `Id` of a `Problem` (`this.submissionsData.GetAllByProblem(id)`) in another case - the `Id` of a `Test` (`this.testData.GetProblemIdByTestId(id)`). The second parameter is there only to control what the firs one will mean..
It should be 2 different public methods.. internally you can still reuse logic, but for the caller it should be clear what the method does and what parameters to provide. If the method is in `ProblemsBusinessService` and is called `RetestById` - then it's a must that the `Id` is only the `Id` of the Problem.

File: Services/Administration/OJS.Services.Administration.Data/Implementations/TestsDataService.cs [hunk 2]
```diff
@@ -30,5 +30,10 @@ public async Task DeleteByProblem(int problemId)
             this.Delete(t => t.ProblemId == problemId);
             await this.SaveChanges();
         }
+
+        public async Task<int> GetProblemIdByTestId(int testId)
+            => await this.GetByIdQuery(testId)
+                .Select(t => t.ProblemId)
+                .FirstAsync();
```
Comment (gogo4ds) [inline Services/Administration/OJS.Services.Administration.Data/Implementations/TestsDataService.cs:37]: FirstAsync will throw exception if there is no test with the provided Id. In DataServices we tend to return null or default if the value is missing and let the caller decide what to do, instead of throwing exception.

File: Services/Common/OJS.Services.Common.Data/Implementations/SubmissionsForProcessingCommonDataService.cs [hunk 3]
```diff
@@ -57,18 +58,15 @@ public async Task Add(int submissionId)
         };
 
         await this.Add(submissionForProcessing);
+
+        return submissionForProcessing;
     }
 
-    public async Task<SubmissionForProcessing> AddOrUpdate(int submissionId)
+    public async Task<SubmissionForProcessing> CreateIfNotExists(int submissionId)
```
Comment (gogo4ds) [inline Services/Common/OJS.Services.Common.Data/Implementations/SubmissionsForProcessingCommonDataService.cs:65]: Why did you change the name of the method? Doesn't it do exactly Adding or Updating, and not just Creating if does not exist?

File: Services/Common/OJS.Services.Common.Data/Implementations/SubmissionsForProcessingCommonDataService.cs [hunk 4]
```diff
@@ -80,15 +78,23 @@ public async Task AddOrUpdateBySubmissionIds(ICollection<int> submissionIds)
...
             {
                 SubmissionId = sId,
+                Processed = false,
+                Processing = false,
             });
 
-        using var scope = TransactionsHelper.CreateTransactionScope();
+        using var scope = TransactionsHelper.CreateTransactionScope(
+            isolationLevel: IsolationLevel.RepeatableRead,
+            asyncFlowOption: TransactionScopeAsyncFlowOption.Enabled);
```
Comment (gogo4ds) [inline Services/Common/OJS.Services.Common.Data/Implementations/SubmissionsForProcessingCommonDataService.cs:87]: Why you need these?

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Services/Common/OJS.Services.Common.Data/Implementations/SubmissionsForProcessingCommonDataService.cs (hunk 4 above)
Comment (PetarNeshkov) [reply Services/Common/OJS.Services.Common.Data/Implementations/SubmissionsForProcessingCommonDataService.cs:87]: It is due to the following problem: "A TransactionScope must be disposed on the same thread that it was created.".   asyncFlowOption: TransactionScopeAsyncFlowOption.Enabled) fixes this.

File: Services/Common/OJS.Services.Common.Data/Implementations/SubmissionsForProcessingCommonDataService.cs (hunk 3 above)
Comment (PetarNeshkov) [reply Services/Common/OJS.Services.Common.Data/Implementations/SubmissionsForProcessingCommonDataService.cs:65]: This logic was modified with pr:  https://github.com/SoftUni-Internal/OpenJudgeSystem/pull/819. I am reverting the previous name.

**Review by gogo4ds — APPROVED** [review]

File: Services/Administration/OJS.Services.Administration.Business/Implementations/ProblemsBusinessService.cs [hunk 5]
```diff
@@ -178,6 +168,35 @@ public async Task<bool> UserHasProblemPermissions(int problemId, string? userId,
         public Task ReevaluateProblemsOrder(int contestId, Problem problem)
             => this.problemGroupsBusiness.ReevaluateProblemsAndProblemGroupsOrder(contestId, problem.ProblemGroup);
 
+        private async Task SetSubmissionsForProcessing(IQueryable<Submission> submissionsQueryable, int problemId)
```
Comment (shsimeonova) [inline Services/Administration/OJS.Services.Administration.Business/Implementations/ProblemsBusinessService.cs:171]: You can pass already mapped submissions to SubmissionServiceModel

**Review by shsimeonova — CHANGES_REQUESTED** [review]

File: Services/Administration/OJS.Services.Administration.Business/Extensions/AdminActionContextExtensions.cs [hunk 6]
```diff
@@ -12,8 +12,17 @@ namespace OJS.Services.Administration.Business.Extensions;
...
+    public static string? GetFormValue(this AdminActionContext actionContext, AdditionalFormFields field)
+    {
+        try
+        {
+            return actionContext.EntityDict[field.ToString()];
+        }
+        catch (KeyNotFoundException)
+        {
+            return null;
+        }
```
Comment (gogo4ds) [inline Services/Administration/OJS.Services.Administration.Business/Extensions/AdminActionContextExtensions.cs:24]: Can you use `GetValueOrDefault` extension method on the Dictionary, or use `HasKey`, instead of catching exception?

**Review by gogo4ds — DISMISSED** [review]

**Review by gogo4ds — DISMISSED** [review]

File: Services/Common/OJS.Services.Common/Implementations/PublisherService.cs [hunk 7]
```diff
@@ -13,4 +14,8 @@ public PublisherService(IBus bus)
     public Task Publish<T>(T obj)
         where T : class
         => this.bus.Publish(obj);
+
+    public Task PublishMultiple<T>(IEnumerable<T> objs)
```
Comment (gogo4ds) [inline Services/Common/OJS.Services.Common/Implementations/PublisherService.cs:18]: Rename to `PublishBatch` as your method is just a wrapper to the `PublishBatch` from MassTransit

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Services/Common/OJS.Services.Common.Models/Submissions/ExecutionContext/SubmissionServiceModel.cs [hunk 8]
```diff
@@ -71,7 +71,7 @@ public void RegisterMappings(IProfileExpression configuration)
                     opt => opt.MapFrom(s => s.SubmissionType!.ExecutionStrategyType))
                 .ForMember(
                     d => d.ExecutionType,
-                    opt => opt.MapFrom(s => ExecutionType.TestsExecution))
+                    opt => opt.MapFrom(s => Workers.Common.Models.ExecutionType.TestsExecution))
```
[the remark this answers is not in the document]
Comment (PetarNeshkov) [reply Services/Common/OJS.Services.Common.Models/Submissions/ExecutionContext/SubmissionServiceModel.cs:74]: It was causing an error with old changes in the worker. At the moment is working appropriately.

File: Servers/Administration/OJS.Servers.Administration/Controllers/TestsController.cs [hunk 9]
```diff
@@ -276,10 +286,48 @@ await this.problemsValidationHelper
...
+
+            await this.problemsBusiness.RetestById(problemId);
+        }
+        else
+        {
+            var problem = await this.testsData.GetProblemById(newTest.Id);
+
+            if (problem != null)
+            {
+                var submissionIds = problem.Submissions.Select(s => s.Id);
```
[the remark this answers is not in the document]
Comment (PetarNeshkov) [reply Servers/Administration/OJS.Servers.Administration/Controllers/TestsController.cs:322]: It does. Unfortunately, I need it, because of the batching.

[the remark this answers is not in the document]
Comment (gogo4ds) [reply Servers/Administration/OJS.Servers.Administration/Controllers/TestsController.cs:320]: You should get directly the `submissionIds`, no need to make 2 requests

File: Services/Common/OJS.Services.Common/Implementations/PublisherService.cs [hunk 10]
```diff
@@ -13,4 +14,8 @@ public PublisherService(IBus bus)
     public Task Publish<T>(T obj)
         where T : class
         => this.bus.Publish(obj);
+
+    public Task Publish<T>(IEnumerable<T> objs)
+        where T : class
+        => this.bus.Publish(objs);
```
Comment (gogo4ds) [inline Services/Common/OJS.Services.Common/Implementations/PublisherService.cs:20]: Why you ditched the `PublishBatch` method? What this does, isn't it the same as the method above?

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Servers/Administration/OJS.Servers.Administration/Controllers/TestsController.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (PetarNeshkov) [reply Servers/Administration/OJS.Servers.Administration/Controllers/TestsController.cs]: Modified.

File: Services/Common/OJS.Services.Common/Implementations/PublisherService.cs (hunk 10 above)
Comment (PetarNeshkov) [reply Services/Common/OJS.Services.Common/Implementations/PublisherService.cs]: Reverted changes and fixed the naming convention.

File: Services/Common/OJS.Services.Common.Models/Submissions/ExecutionContext/SubmissionServiceModel.cs (hunk 8 above)
Comment (shsimeonova) [inline Services/Common/OJS.Services.Common.Models/Submissions/ExecutionContext/SubmissionServiceModel.cs:74]: Unnessessary prefix here

**Review by shsimeonova — CHANGES_REQUESTED** [review]

File: Services/Common/OJS.Services.Common.Models/Submissions/ExecutionContext/SubmissionServiceModel.cs (hunk 8 above)
Comment (PetarNeshkov) [reply Services/Common/OJS.Services.Common.Models/Submissions/ExecutionContext/SubmissionServiceModel.cs:74]: It is needed. Otherwise it is causing an error in rider.

File: Services/Administration/OJS.Services.Administration.Data/Implementations/SubmissionsDataService.cs [hunk 11]
```diff
@@ -38,12 +39,12 @@ public IQueryable<Submission> GetAllByProblemAndParticipant(int problemId, int p
 
         public IQueryable<Submission> GetAllFromContestsByLecturer(string lecturerId)
             => this.DbSet
-                .Include(s => s.Problem!.ProblemGroup.Contest.LecturersInContests)
-                .Include(s => s.Problem!.ProblemGroup.Contest.Category!.LecturersInContestCategories)
+                .Include(s => s.Problem.ProblemGroup.Contest.LecturersInContests)
+                .Include(s => s.Problem.ProblemGroup.Contest.Category!.LecturersInContestCategories)
```
Comment (gogo4ds) [inline Services/Administration/OJS.Services.Administration.Data/Implementations/SubmissionsDataService.cs:43]: Use `ThenInclude` instead of chaining in a single Include

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Services/Common/OJS.Services.Common.Models/Submissions/ExecutionContext/SubmissionServiceModel.cs (hunk 8 above)
Comment (shsimeonova) [reply Services/Common/OJS.Services.Common.Models/Submissions/ExecutionContext/SubmissionServiceModel.cs:74]: What is the error? 
image
It seems like it can be removed.

File: Services/Administration/OJS.Services.Administration.Data/Implementations/SubmissionsDataService.cs (hunk 11 above)
Comment (PetarNeshkov) [reply Services/Administration/OJS.Services.Administration.Data/Implementations/SubmissionsDataService.cs]: The following code has no impact on the created PR. Reverting my changes.

File: Services/Common/OJS.Services.Common.Models/Submissions/ExecutionContext/SubmissionServiceModel.cs (hunk 8 above)
Comment (PetarNeshkov) [reply Services/Common/OJS.Services.Common.Models/Submissions/ExecutionContext/SubmissionServiceModel.cs:74]: The problem seems to be fixed. Removed the unused namespace.

**Review by shsimeonova — APPROVED** [review]

**Review by gogo4ds — APPROVED** [review]

**Review by shsimeonova — APPROVED** [review]

**Review by gogo4ds — DISMISSED** [review]

**Review by gogo4ds — APPROVED** [review]
